### PR TITLE
Provide instructions for resource change notifications in tools

### PR DESCRIPTION
**Pull request by betalars**
I have found this quite counter-intuitive, and was irritated by there being no guidance around this, so I wanted to create this enhancement.

Please double-check my c# Code, I do not use it personally and there will probably be mistakes.

### CODE REVIEW

**Comment by betalars** [comment]
Also: I know this is not strictly adhering to the guidelines when it comes to giving easy-to-understand examples, but honestly this is kind of an advanced issue so I feel there is a value in keeping it short?

But feel free to suggest a different framing and I will update it.

Most examples of resources I could think of would add a lot of complexity to the explanation because using a resource in a tool does not make sense that often.

File: tutorials/plugins/running_code_in_the_editor.rst
Comment (AThousandShips) [inline tutorials/plugins/running_code_in_the_editor.rst:261]:
```suggestion
    @export var resource: MyResource:
```

File: tutorials/plugins/running_code_in_the_editor.rst
Comment (AThousandShips) [inline tutorials/plugins/running_code_in_the_editor.rst:313]:
```suggestion
```

File: tutorials/plugins/running_code_in_the_editor.rst
Comment (AThousandShips) [inline tutorials/plugins/running_code_in_the_editor.rst:304]:
```suggestion

    # Make Your Resource a tool.
```

File: tutorials/plugins/running_code_in_the_editor.rst
Comment (AThousandShips) [inline tutorials/plugins/running_code_in_the_editor.rst:353]:
```suggestion
    @export var resource: MyResource:
```

File: tutorials/plugins/running_code_in_the_editor.rst
Comment (AThousandShips) [inline tutorials/plugins/running_code_in_the_editor.rst:328]:
```suggestion
        // Declare a signal.
```

File: tutorials/plugins/running_code_in_the_editor.rst
Comment (AThousandShips) [inline tutorials/plugins/running_code_in_the_editor.rst:339]:
```suggestion
                // Emit a signal when the property is changed.
```

File: tutorials/plugins/running_code_in_the_editor.rst
Comment (AThousandShips) [inline tutorials/plugins/running_code_in_the_editor.rst:378]:
```suggestion
                // Connect the signal you just declared as soon as a new resource is being added.
```

File: tutorials/plugins/running_code_in_the_editor.rst
Comment (AThousandShips) [inline tutorials/plugins/running_code_in_the_editor.rst:267]:
```suggestion
    # This will only be called when you create, delete, or paste a resource.
    # You will not get an update when tweaking properties of it.
```

File: tutorials/plugins/running_code_in_the_editor.rst
Comment (AThousandShips) [inline tutorials/plugins/running_code_in_the_editor.rst:293]:
```suggestion
    // This will only be called when you create, delete, or paste a resource.
    // You will not get an update when tweaking properties of it.
```

File: tutorials/plugins/running_code_in_the_editor.rst
Comment (AThousandShips) [inline tutorials/plugins/running_code_in_the_editor.rst:296]:
```suggestion
        GD.Print("My resource was set!");
```

File: tutorials/plugins/running_code_in_the_editor.rst
Comment (AThousandShips) [inline tutorials/plugins/running_code_in_the_editor.rst:386]:
```suggestion
        GD.Print("My Resource just updated!");
```

File: tutorials/plugins/running_code_in_the_editor.rst
Comment (AThousandShips) [inline tutorials/plugins/running_code_in_the_editor.rst:316]:
```suggestion
            # Emit a signal when the property is changed.
```

File: tutorials/plugins/running_code_in_the_editor.rst
Comment (AThousandShips) [inline tutorials/plugins/running_code_in_the_editor.rst:249]:
```suggestion
Getting notified when resources change
--------------------------------------

```

File: tutorials/plugins/running_code_in_the_editor.rst
Comment (AThousandShips) [inline tutorials/plugins/running_code_in_the_editor.rst:252]:
```suggestion
Some times you want your tool to use a resource. However, when you change an
property of that resource in the editor, the ``set()`` method of your tool will
not be called.
```

File: tutorials/plugins/running_code_in_the_editor.rst
Comment (AThousandShips) [inline tutorials/plugins/running_code_in_the_editor.rst:345]:
```suggestion
You then want to connect the signal when a new resource is set:

```

File: tutorials/plugins/running_code_in_the_editor.rst
Comment (AThousandShips) [inline tutorials/plugins/running_code_in_the_editor.rst:299]:
```suggestion
To get around this problem you first have to make your resource a tool and make it
```

File: tutorials/plugins/running_code_in_the_editor.rst
Comment (AThousandShips) [inline tutorials/plugins/running_code_in_the_editor.rst:300]:
```suggestion
emit a signal whenever a property is set:
```

File: tutorials/plugins/running_code_in_the_editor.rst
Comment (betalars) [reply tutorials/plugins/running_code_in_the_editor.rst:249]: Hm I personally like to capitalise Keywords, and in this case it is referring to a class. But I have just looked around the wiki and that seems to not be common practce.

File: tutorials/plugins/running_code_in_the_editor.rst
Comment (raulsntos) [inline tutorials/plugins/running_code_in_the_editor.rst:288]:
```suggestion
                OnResourceSet();
```

**Comment by betalars** [comment]
Thanks @AThousandShips !

File: tutorials/plugins/running_code_in_the_editor.rst
Comment (raulsntos) [inline tutorials/plugins/running_code_in_the_editor.rst:279]:
```suggestion
        private MyResource _resource;
```

**Comment by AThousandShips** [comment]
You missed several ones, please check all

File: tutorials/plugins/running_code_in_the_editor.rst
Comment (raulsntos) [inline tutorials/plugins/running_code_in_the_editor.rst:371]:
```suggestion
        private MyResource _resource;
```

File: tutorials/plugins/running_code_in_the_editor.rst
Comment (raulsntos) [inline tutorials/plugins/running_code_in_the_editor.rst:359]: If the previous `resource` value was not null, then you also need to disconnect from its signal.

File: tutorials/plugins/running_code_in_the_editor.rst
Comment (raulsntos) [inline tutorials/plugins/running_code_in_the_editor.rst:312]: Instead of declaring and emitting a custom signal, we may want to direct users towards the `changed` signal and the `emit_changed` method.

**Comment by betalars** [comment]
> You missed several ones, please check all

Yea I just realized that as well while checking the diff. I did not expect github to collapse unresolved threads  and not resolved ones.

File: tutorials/plugins/running_code_in_the_editor.rst
Comment (raulsntos) [inline tutorials/plugins/running_code_in_the_editor.rst:388]: To match the GDScript example.

```suggestion
        GD.Print("My resource just changed!");
```

**Comment by raulsntos** [comment]
Also note that you can use the `Add suggestion to batch` button from the `Files changed` tab instead of applying the suggestions one by one.

File: tutorials/plugins/running_code_in_the_editor.rst
Comment (betalars) [reply tutorials/plugins/running_code_in_the_editor.rst:288]: My arch enemy, the semicolon.

File: tutorials/plugins/running_code_in_the_editor.rst
Comment (betalars) [reply tutorials/plugins/running_code_in_the_editor.rst:312]: Okay, I did not find that signal, it seems really useful.

File: tutorials/plugins/running_code_in_the_editor.rst
Comment (betalars) [reply tutorials/plugins/running_code_in_the_editor.rst:359]: okay, will do.

**Comment by betalars** [comment]
Oh no, I think starting a review was the wrong move ... now everything is pending and I cannot merge it myself.

File: tutorials/plugins/running_code_in_the_editor.rst
Comment (betalars) [reply tutorials/plugins/running_code_in_the_editor.rst:359]: Have done it. I once again cannot commit it myself, because the review is pending. Sorry about that :c

**Review by betalars — COMMENTED** [review]
Applying suggestion to switch from a costum signal to changed.

**Comment by betalars** [comment]
Oh I see how github works.

File: tutorials/plugins/running_code_in_the_editor.rst
Comment (AThousandShips) [inline tutorials/plugins/running_code_in_the_editor.rst:359]:
```suggestion
            resource.property_updated.connect(_on_resource_updated)
```

File: tutorials/plugins/running_code_in_the_editor.rst
Comment (AThousandShips) [inline tutorials/plugins/running_code_in_the_editor.rst:352]:
```suggestion
            resource.changed.connect(_on_resource_changed)
```

File: tutorials/plugins/running_code_in_the_editor.rst
Comment (betalars) [reply tutorials/plugins/running_code_in_the_editor.rst]: Today on more things I would have liked to know earlier.

File: tutorials/plugins/running_code_in_the_editor.rst
Comment (betalars) [inline tutorials/plugins/running_code_in_the_editor.rst:383]: Recommending users to also disconnect the signals. I think it is fair to put it last and not include it above, because it will rarely cause issues (just some unneeded updates) when overlooked and would mix up the order of explanation otherwise.
```suggestion
Lastly, you need to disconnect the signal if the previous resource was not ``null.``
.. tabs::
 .. code-tab:: gdscript GDScript

    @export var resource: MyResource:
        set(new_resource):
            if resource != null: resource.changed.disconnect(_on_resource_changed)
            resource = new_resource
            resource.changed.connect(_on_resource_changed)
.. code-tab:: csharp
    [Export]
    public MyResource Resource
    {
        get => _resource;
        set
        {
            // Disconnect the signal if the previous resource was not null
            if (_resource != null) { _resource.Changed -= OnResourceChanged; }
            _resource = value;
            _resource.Changed += OnResourceChanged;
        }
    }

```

File: tutorials/plugins/running_code_in_the_editor.rst
Comment (betalars) [reply tutorials/plugins/running_code_in_the_editor.rst]: I added a new section to explain this.

File: tutorials/plugins/running_code_in_the_editor.rst
Comment (AThousandShips) [inline tutorials/plugins/running_code_in_the_editor.rst:395]:
```suggestion

.. code-tab:: csharp

```

File: tutorials/plugins/running_code_in_the_editor.rst
Comment (AThousandShips) [inline tutorials/plugins/running_code_in_the_editor.rst:392]:
```suggestion
            if resource != null:
                resource.changed.disconnect(_on_resource_changed)
```

File: tutorials/plugins/running_code_in_the_editor.rst
Comment (AThousandShips) [inline tutorials/plugins/running_code_in_the_editor.rst:385]:
```suggestion
would cause unneeded updates.

```

File: tutorials/plugins/running_code_in_the_editor.rst
Comment (AThousandShips) [inline tutorials/plugins/running_code_in_the_editor.rst:405]:
```suggestion
            if (_resource != null)
            {
                _resource.Changed -= OnResourceChanged;
            }
```
Same here IMO, @raulsntos

**Review by AThousandShips — APPROVED** [review]
Stylewise this look great!

File: tutorials/plugins/running_code_in_the_editor.rst
Comment (AThousandShips) [inline tutorials/plugins/running_code_in_the_editor.rst:398]:
```suggestion
 .. code-tab:: csharp
```

File: tutorials/plugins/running_code_in_the_editor.rst
Comment (raulsntos) [inline tutorials/plugins/running_code_in_the_editor.rst:334]: There's a dedicated method for this in `Resource`.

```suggestion
                EmitChanged();
```

**Review by raulsntos — APPROVED** [review]

**Comment by AThousandShips** [comment]
Thanks! And congratulations on your first merged contribution! :tada:

**Comment by betalars** [comment]
> Thanks! And congratulations on your first merged contribution! 🎉

I think I did another documentation thing before, but this is the first merge with code :D

**Comment by AThousandShips** [comment]
It wasn't registered at least, at least not on here

**Comment by betalars** [comment]
> Lastly, you should to disconnect
Oh no, we missed a typo.
I really need to get into a habit of letting things sit for a day and then checking again because I often restructure sentences and then make gramatical errors because of that.

File: tutorials/plugins/running_code_in_the_editor.rst
Comment (betalars) [inline tutorials/plugins/running_code_in_the_editor.rst:384]:
```suggestion
Lastly, remember to disconnect the signal as the old resource being used and changed somewhere else
```

Comment (betalars) [reply tutorials/plugins/running_code_in_the_editor.rst:384]: okay I am afraid this is going to need a new PR

Comment (AThousandShips) [reply tutorials/plugins/running_code_in_the_editor.rst:384]: Yes you can't update something already merged, but no worries, missed it too

**Comment by mhilbrunner** [comment]
Cherry-picked to 4.2 in https://github.com/godotengine/godot-docs/pull/9648.

**Pull request merged**
